### PR TITLE
VBLOCKS-2180 | AudioProcessor implementation

### PR DESCRIPTION
**Pull request by charliesantos**
**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

## Pull Request Details

AudioProcessor implementation. See https://issues.corp.twilio.com/browse/VBLOCKS-2180

### Before review
* [ ] Updated CHANGELOG.md if necessary
* [ ] Added unit tests if necessary
* [ ] Updated affected documentation
* [ ] Verified locally with `npm test`
* [x] Manually sanity tested running locally
* [x] Ready for review

### Before merge
* [ ] Got one or more +1s
* [ ] Squashed erroneous commits if necessary
* [ ] Re-tested if necessary

### CODE REVIEW

File: lib/twilio/audiohelper.ts [hunk 1]
```diff
@@ -302,6 +320,39 @@ class AudioHelper extends EventEmitter {
     this._isPollingInputVolume = false;
   }
 
+  /**
+   * Call getUserMedia with specified constraints
+   * @private
+   */
+  _openDefaultDeviceWithConstraints(constraints: MediaStreamConstraints): Promise<MediaStream> {
```
Comment (manjeshbhargav) [inline lib/twilio/audiohelper.ts:327]: Can we use `async/await` here?

File: lib/twilio/audiohelper.ts [hunk 2]
```diff
@@ -516,6 +588,20 @@ class AudioHelper extends EventEmitter {
     });
   }
 
+  /**
+   * Route input stream to the processor if it exists
+   */
+  private _maybeCreateProcessedStream(stream: MediaStream): Promise<MediaStream> {
```
Comment (manjeshbhargav) [inline lib/twilio/audiohelper.ts:594]: `async/await`?

File: lib/twilio/audiohelper.ts [hunk 3]
```diff
@@ -569,13 +656,35 @@ class AudioHelper extends EventEmitter {
...
     }
 
-    this._inputStream = stream;
+    this._selectedInputDeviceStream = stream;
+  }
+
+  /**
+   * Restart the active streams
+   */
+  private _restartStreams(): Promise<void> {
```
Comment (manjeshbhargav) [inline lib/twilio/audiohelper.ts:673]: `async/await`?

File: lib/twilio/audiohelper.ts [hunk 4]
```diff
@@ -595,24 +704,37 @@ class AudioHelper extends EventEmitter {
...
+    this._stopDefaultInputDeviceStream();
+
     const constraints = { audio: Object.assign({ deviceId: { exact: deviceId } }, this.audioConstraints) };
-    return this._getUserMedia(constraints).then((stream: MediaStream) => {
-      return this._onActiveInputChanged(stream).then(() => {
-        this._replaceStream(stream);
-        this._inputDevice = device;
-        this._maybeStartPollingVolume();
+    this._log.debug('setInputDevice: getting new tracks.');
+    return this._getUserMedia(constraints).then((originalStream: MediaStream) => {
```
Comment (manjeshbhargav) [inline lib/twilio/audiohelper.ts:727]: `async/await`?

File: lib/twilio/audiohelper.ts [hunk 5]
```diff
@@ -595,24 +704,37 @@ class AudioHelper extends EventEmitter {
...
+
+      this._destroyProcessedStream();
+
+      return this._maybeCreateProcessedStream(originalStream).then((newStream) => {
```
Comment (manjeshbhargav) [inline lib/twilio/audiohelper.ts:731]: `async/await`?

File: lib/twilio/audiohelper.ts [hunk 6]
```diff
@@ -595,24 +704,37 @@ class AudioHelper extends EventEmitter {
...
+        this._log.debug('setInputDevice: invoking _onActiveInputChanged.');
+        return this._onActiveInputChanged(newStream).then(() => {
```
Comment (manjeshbhargav) [inline lib/twilio/audiohelper.ts:733]: `async/await`?

**Comment by charliesantos** [comment]
@manjeshbhargav I considered using async/await however, it won't be consistent with the rest of the file since it's still using promise style. Do you think it still make sense to do it?

**Comment by manjeshbhargav** [comment]
@charliesantos At least we can start making the changes in the files that are modified.

**Review by manjeshbhargav — APPROVED** [review]

**Pull request merged**